### ISSUE TITLE
Proposal Batch-Crossref importer

### ISSUE THREAD

**Issue by etiennedi** [creativesoftwarefdn/weaviate#781]
tl;dr: The following three posts show how my ideas for this endpoint evolved. If you're only interested in my latest proposal, please see https://github.com/creativesoftwarefdn/weaviate/issues/781#issuecomment-473684432. Nevertheless question "2" from this post is still relevant.

## Problem
* We know from our scale tests (#713) that a one-by-one import is slow if the scale is very large.
* We were able to improve the (Janusgraph) import speed considerably by rewriting the batch importers to send a single request to the HTTP server countaining 25-50 imports at once
* The current batch importers only support primitive properties
* Because it is a requirement for a cross-ref that both the source and target thing/action are already present we must import cross-refs separate from the things/actions * 

 *= I'm not sure if this is always true, for example if we only have Cities and Persons and only a one-way relationship `Person -> LivesIn -> City` it might (!) be possible to first import all Cities and then import all Persons directly with the cross-ref. However, this will definitely not work if we have something like `Person -> knows -> Person`. In this case we must decouple importing Vertices and Edges.

## Proposal: A batch importer specific to cross-refs
Route: `POST /batching/<thingsOrActions>/properties/<cross-ref-prop-name>`
Note: I don't like the resource `properties`, as it suggests that you can add all properties this way. However, I chose it so it is consistent with `POST /things/properties/<cross-ref-prop-name>` which can also only be used to add cross-refs. We could also rename this to `references` here and then call it this way everywhere.
Body:
```json
[
  { 
    "from": "thing-or-action-uuid",
    "$cref": "weaviate://localhost/things/thing-uuid-of-target"
  },
  { 
    "from": "could-be-the-same-or-another-thing-or-action-uuid",
    "$cref": "weaviate://localhost/things/thing-uuid-of-another-target"
  },
]
```
Behavior for different cardinalities: The same. The batch importer assumes that either no cross-refs are present yet or that the cardinality is set to `many`. Therefore we can add one in either case. If the cardinality is set to `atMostOne` and there is already one present, this particular class kind would fail the validation.

## Feedback requested
@bobvanluijt @laura-ham @JeroenStravers1 @dljcollette

1. How do you like the proposed cross-ref batch importer as outlined above?
2. Should we name it `/batching/things/properties/<cross-ref-prop-name>` so it's consistent with the other endpoint. Or should we name it `/batching/things/references/<cross-ref-prop-name>`(which in my opinion is more correct) and then also change the existing routes from `properties -> references`?

**Comment by etiennedi** [comment]
Update: I already saw a mistake in the above proposal. We can't reference a property without specifying the class. However, the user might want to make references from different class types in a single batch. So I think we need to move that into the the request body.

## Updated Proposal
Route `POST /batching/properties` (or alternatively `POST /batching/references/)

```
[
  { 
    "from": "a-thing-id-of-a-city",
    "property": "InCountry",
    "$cref": "weaviate://localhost/things/thing-uuid-of-target"
  },
  { 
    "from": "a-thing-id-of-a-person",
    "property": "LivesIn",
    "$cref": "weaviate://localhost/things/thing-uuid-of-another-target"
  },
]
```

**Comment by etiennedi** [comment]
Based on the last proposal we could also try to write this more concise like so:
```json
{ 
    "from": "weaviate://localhost/things/thing-uuid-of-a-City/InCountry",
    "to": "weaviate://localhost/things/thing-uuid-of-target"
}
```

I think I might actually prefer this one. The URL structure with a new `/<property>` section is new I believe, we've never had that before. But I think it makes sense.

**Comment by laura-ham** [comment]
Hi @etiennedi,

Question: Do we need to specify the `kind` of the `from` class? Because you give the uuid, so whether it is a `Thing` or an `Action` doesn't really matter and can't be changed.

Then, what is exactly what you want to do with this? Because your initial proposal and the first comment achieve something different:
- Adding references with the **same** property name
    - As your initial proposal
    - Can be done from and to multiple different classes, but prop name should be the same. That doesn't really make sense to me
- Adding references, the **don't have to have anything in common**
    - Like in our first comment.
    - pro: can add many different references in one batch
    - con: If you want to add many different references from one class, you need to specify the `from` part a lot of times
Your second comment would indeed make more sense then, but then the `/<property>` should be in the URL again then, which makes it the goal the same as your initial proposal again.

And I like `POST /batching/references/` better then `POST /batching/properties/, because in the proposal there is indeed only the option to add references, and not all type of properties..


But why do we want to 'sort' or categorize the batch based on the property name, and not the `from` class? Having the `/<property>` in the URL and not the `thingId` lets you specify something 'in' the Thing, without having specified the 'Thing' yet, it's kind of skipping one layer or so.. Also, is it likely that there are things of various classes with the same property name you want to add a reference to? I don't think so right now?

So what about adding a batch of properties per class like:

`POST /batching/things/<thingId>`
```json
[
  { 
    "property": "InCountry",
    "$cref": "weaviate://localhost/things/thing-uuid-of-target"
  }
]
```

or:

`POST /batching/things/<thingId>/properties/<propertyName>`
```json
[
  { 
    "$cref": "weaviate://localhost/things/thing-uuid-of-target"
  }
]
```

But maybe this only works when the cardinality of the reference is `hasMany`, or the request should be `PUT` (which is not an option because it replaces the current content)..

**Comment by bobvanluijt** [comment]
The proposal makes complete sense. The only remark I have is regarding the nomenclature, there I’m with Laura, I think `references` is better suited.

**Comment by etiennedi** [comment]
@bobvanluijt 

> The only remark I have is regarding the nomenclature, there I’m with Laura, I think references is better suited.

Fully agree, do you also agree that we should change the existing routes like
`/things/properties/<cross-ref-prop-name>` to `/things/references/<cross-ref-prop-name>` since they are essentially the same thing, but currently called `properties`?

**Comment by bobvanluijt** [comment]
Good point, yes, I found it confusing while working on the docs because they _only_ serve references and not properties in general.

**Comment by etiennedi** [comment]
@laura-ham 

Thanks for your comments and questions. Let me try to address them one by one.

> Question: Do we need to specify the kind of the from class? Because you give the uuid, so whether it is a Thing or an Action doesn't really matter and can't be changed.

It's not a hard-requirement. So far everything we have to access a specific class item does require specifying the kind though. For example `GET /things/<uuid>` or setting a cross-ref (without batch) where we specify something like `weaviate://localhost/things/<uuid>`. So I'd prefer if we also included the kind here for two reasons:

1. It's consistent, even though uuids are in fact unique and enough to identify a class, currently you always need to be aware if you're aksing for a `Thing` or `Action`. I could also see this be more relevant if we ever had more Kinds.
2. With the database connector abstraction model, we need to try and support as many dbs as possible. This also means they all have different indexing strategies. With more information in the query we could potentially cut the size of an index in half. I.e. instead of having 1bn classes indexed by uuid, we could have 500m indexed by uuid in a `Things` index and the same for `Actions`. This should in theory make look-ups more efficient. (The effect of that varies from connector to connector, but I think it's better to be safe in this case since we don't know which connectors will come, but explicitly want to support many).

> Then, what is exactly what you want to do with this? Because your initial proposal and the first comment achieve something different

Very good point. My comments evolved a bit as my ideas did and it now seems as if the target different goals. So, to re-iterate: The goal of this batching endpoint is not to make it easier for the user to group references by any logical units. The only goal is to import as many references (that might have something in common, but could also be completely independent) in a single request. This is because we've noticed (with Janusgraph, but I assume it applies for others), that it's much slower to do 10 imports of X, rather than a single import of 10X.

> Can be done from and to multiple different classes, but prop name should be the same. That doesn't really make sense to me

Agree, this was the major flaw in my first proposal. Somehow I though it would make sense to group it by ref-prop. Not sure why, but completely missed the fact that all `from` class instances would then have to be of the same class.

> pro: can add many different references in one batch
con: If you want to add many different references from one class, you need to specify the from part a lot of times

The pro is what I'm trying to achieve. The con is less of an issue I think, since the batching imports will almost always be fed by a script anyway, so it shouldn't be a additional effort to repeat the from class over and over again. Request size should also be no issue (based on my initial tests with importing without cross-refs). But in exchange we get the ability to import completely independent cross-refs.

> But why do we want to 'sort' or categorize the batch based on the property name, and not the from class? 

I don't. I only wanted to do that in the first proposal, but then realized that doesn't make sense (as you did too).

The URL that the user would sent the batch request to in the "final" proposal would simply be `POST /batching/references/` (I've already renamed it to `references` here since we're all in agreement that that's the better term).

Did maybe this sentence of mine confuse you?
> The URL structure with a new /<property> section in the beacon url is new. I believe, we've never had that before. But I think it makes sense.

Here I didn't mean the URL that the user send the request to, but in the body the `from` field which is in turn a URL structure, like `weaviate://localhost/things/<uuid>/<RefPropName>`.

Your final two proposals make sense from a UX perspective, but unfortunately don't solve the "I want to be able to group 50 cross-refs into a single import that might have nothing in common" problem. I think we can use the existing `PATCH /things/uuid` endpoint to add multiple cross-refs from a single class already, so we don't need to add this through patching.

I hope my comments helped clarify some of the confusion from before. Keeping all that in mind, would you agree the following proposal is the best fit for the requirements?

`POST /batching/references` with the following body:
```json
[
  { 
    "from": "weaviate://localhost/things/thing-uuid-of-a-City/InCountry",
    "to": "weaviate://localhost/things/thing-uuid-of-target"
  },
  { 
    "from": "weaviate://localhost/things/thing-uuid-of-a-Person/LivesIn",
    "to": "weaviate://localhost/things/thing-uuid-of-target"
  },
]
```

**Comment by laura-ham** [comment]
Hi @etiennedi ,

Thanks, this totally makes sense!

I now also see that my final proposal does indeed not solve the problem.

Your last proposal, for `POST /batching/references` looks good! :)

**Comment by etiennedi** [comment]
Closing in favor of #788 as the proposal is considered accepted and now we only need to track the implementation.
